File: scripts/sync.ts
import fs from 'fs-extra';
import path from 'path';
import axios from 'axios';

interface UpstreamSource {
  name: string;
  url: string;
  category: 'proxy' | 'direct' | 'reject';
  target: string;
  enabled: boolean;
}

const findRoot = () => {
  let curr = process.cwd();
  while (curr !== path.parse(curr).root) {
    if (fs.existsSync(path.join(curr, '.git'))) return curr;
    curr = path.dirname(curr);
  }
  return '.';
};

async function main() {
  const root = findRoot();
  const configPath = path.join(root, 'source', 'sources.json');
  const sources: UpstreamSource[] = await fs.readJson(configPath);
  const upstreamDir = path.join(root, 'source', 'upstream');

  console.log('============================================================');
  console.log('  Gins-Rules Upstream Syncer (TS/Bun)');
  console.log('============================================================');

  const mergedResults: Record<string, Record<string, string[]>> = {};

  for (const src of sources) {
    if (!src.enabled) continue;
    console.log(`  Fetching ${src.name}...`);

    try {
      const resp = await axios.get(src.url, { timeout: 30000 });
      const rules = resp.data.split('\n').filter((l: string) => l.trim() && !l.startsWith('#')).map((l: string) => l.trim());
      
      if (!mergedResults[src.category]) mergedResults[src.category] = {};
      if (!mergedResults[src.category][src.target]) mergedResults[src.category][src.target] = [];
      mergedResults[src.category][src.target].push(...rules);
    } catch (err: any) {
      console.error(`  [ERROR] ${src.name}: ${err.message}`);
    }
  }

  for (const [cat, targets] of Object.entries(mergedResults)) {
    for (const [name, rules] of Object.entries(targets)) {
      const outPath = path.join(upstreamDir, cat, `${name}.txt`);
      await fs.ensureDir(path.dirname(outPath));
      await fs.writeFile(outPath, Array.from(new Set(rules)).sort().join('\n') + '\n');
    }
  }

  console.log('  Sync complete!');
}

main();

File: tools/gins-rules-engine/src/sync.rs
use anyhow::Result;
use serde::Deserialize;
use std::collections::HashMap;
use std::path::PathBuf;

#[derive(Debug, Deserialize)]
struct UpstreamSource {
    name: String,
    url: String,
    category: String,
    target: String,
    enabled: bool,
}

pub async fn run(root: &str) -> Result<()> {
    let root = PathBuf::from(root).canonicalize()?;
    let sources_path = root.join("source/sources.json");
    let data = tokio::fs::read(&sources_path).await?;
    let sources: Vec<UpstreamSource> = serde_json::from_slice(&data)?;

    let active: Vec<_> = sources.into_iter().filter(|s| s.enabled).collect();
    println!("📡 Syncing {} active rule sources...", active.len());

    let client = reqwest::Client::builder()
        .user_agent("Gins-Rules-Engine/2.0")
        .build()?;

    let mut collected: HashMap<String, HashMap<String, String>> = HashMap::new();

    let mut join_set = tokio::task::JoinSet::new();
    for src in active {
        let c = client.clone();
        join_set.spawn(async move {
            println!("  [Rule] Downloading {}...", src.name);
            let resp = c.get(&src.url).send().await?.text().await?;
            Ok::<_, anyhow::Error>((src.category, src.target, resp))
        });
    }

    while let Some(result) = join_set.join_next().await {
        if let Ok(Ok((cat, target, text))) = result {
            let cat_map = collected.entry(cat).or_default();
            let entry = cat_map.entry(target).or_default();
            entry.push_str(&text);
            entry.push('\n');
        } else if let Ok(Err(e)) = result {
            eprintln!("  ❌ Error downloading: {}", e);
        }
    }

    for (cat, targets) in collected {
        let dir = root.join("source/upstream").join(&cat);
        tokio::fs::create_dir_all(&dir).await?;
        for (target, content) in targets {
            let file_path = dir.join(format!("{}.txt", target));
            tokio::fs::write(&file_path, content).await?;
            println!("  ✅ Saved {}/{}", cat, target);
        }
    }

    println!("✨ Rule sync completed.");
    Ok(())
}
